bridges/snowbridge/primitives/outbound-queue/src/v2/converter/convert.rs

```diff
@@ -0,0 +1,270 @@
+			beneficiary.unpack(),
+			(0, [AccountKey20 { network, key }])
+				if self.network_matches(network),
+			H160(*key)
+		)
+		.ok_or(BeneficiaryResolutionFailed)?;
+
+		// Make sure there are reserved assets.
+		if enas.is_none() && pnas.is_none() {
+			return Err(NoReserveAssets)
+		}
+
+		let mut commands: Vec<Command> = Vec::new();
+
+		// ENA transfer commands
+		if let Some(enas) = enas {
+			for ena in enas.clone().inner().iter() {
+				// Check the the deposit asset filter matches what was reserved.
+				if !deposit_assets.matches(ena) {
+					return Err(FilterDoesNotConsumeAllAssets)
+				}
+
+				// only fungible asset is allowed
+				let (token, amount) = match ena {
+					Asset { id: AssetId(inner_location), fun: Fungible(amount) } =>
+						match inner_location.unpack() {
+							(0, [AccountKey20 { network, key }])
+								if self.network_matches(network) =>
+								Some((H160(*key), *amount)),
+							_ => None,
+						},
+					_ => None,
+				}
+				.ok_or(AssetResolutionFailed)?;
+
+				// transfer amount must be greater than 0.
+				ensure!(amount > 0, ZeroAssetTransfer);
+
+				commands.push(Command::UnlockNativeToken { token, recipient, amount });
+			}
+		}
+
+		// PNA transfer commands
+		if let Some(pnas) = pnas {
+			ensure!(pnas.len() > 0, NoReserveAssets);
+			for pna in pnas.clone().inner().iter() {
+				// Check the the deposit asset filter matches what was reserved.
+				if !deposit_assets.matches(pna) {
+					return Err(FilterDoesNotConsumeAllAssets)
+				}
+
+				// Only fungible is allowed
+				let (asset_id, amount) = match pna {
+					Asset { id: AssetId(inner_location), fun: Fungible(amount) } =>
+						Some((inner_location.clone(), *amount)),
+					_ => None,
+				}
+				.ok_or(AssetResolutionFailed)?;
+
+				// transfer amount must be greater than 0.
+				ensure!(amount > 0, ZeroAssetTransfer);
+
+				// Ensure PNA already registered
+				let token_id = TokenIdOf::convert_location(&asset_id).ok_or(InvalidAsset)?;
+				let expected_asset_id = ConvertAssetId::convert(&token_id).ok_or(InvalidAsset)?;
+				ensure!(asset_id == expected_asset_id, InvalidAsset);
+
+				commands.push(Command::MintForeignToken { token_id, recipient, amount });
+			}
+		}
+
+		// Transact commands
+		let transact_call = match_expression!(self.peek(), Ok(Transact { call, .. }), call);
+		if let Some(transact_call) = transact_call {
+			let _ = self.next();
+			let transact =
+				TransactInfo::decode_all(&mut transact_call.clone().into_encoded().as_slice())
+					.map_err(|_| TransactDecodeFailed)?;
+			commands.push(Command::CallContract {
+				target: transact.target,
+				data: transact.data,
+				gas_limit: transact.gas_limit,
+				value: transact.value,
+			});
+		}
+
+		// ensure SetTopic exists
+		let topic_id = match_expression!(self.next()?, SetTopic(id), id).ok_or(SetTopicExpected)?;
+
+		let message = Message {
+			id: (*topic_id).into(),
+			origin_location: origin_location.clone(),
+			origin,
```

Comment:
@yrong remind me why we need to include both `origin_location` and `origin`?

Reply:
Agree, it's not required and removed in https://github.com/vgeddes/polkadot-sdk/pull/6/commits/66ac16313aaa593328f8977a88a663edea10e9a7